File: 2025-season6/9주차/최영권/KClosestPointsToOrigin.java
package LinkedList;

import java.util.Arrays;
import java.util.PriorityQueue;

/**
 * 시간복잡도 : O(NlogN) , N은 점의 총 개수
 */
public class KClosestPointsToOrigin {
    public static void main(String[] args) {
        KClosestPointsToOrigin obj = new KClosestPointsToOrigin();
        int[][] x = obj.kClosest(new int[][]{{1, 3}, {-2, 2}}, 1);
        System.out.println(Arrays.deepToString(x));
    }

    public int[][] kClosest(int[][] points, int k) {
        int[][] result = new int[k][2];
        PriorityQueue<Point> pq = new PriorityQueue<>();
        for (int[] point : points) {
            int x = point[0];
            int y = point[1];
            int abs1 = Math.abs(x);
            int abs2 = Math.abs(y);
            int distance = (abs1 * abs1) + (abs2 * abs2);
            Point point1 = new Point(x, y, distance);
            pq.offer(point1);
        }
        for (int i = 0; i < k; i++) {
            Point poll = pq.poll();
            result[i][0] = poll.x;
            result[i][1] = poll.y;
        }
//        System.out.println(pq);
        return result;
    }

    static class Point implements Comparable<Point> {
        int x;
        int y;
        int distance;

        public Point(int x, int y, int distance) {
            this.x = x;
            this.y = y;
            this.distance = distance;
        }

        @Override
        public int compareTo(Point o) {
            return this.distance - o.distance;
        }
    }
}

File: 2025-season6/9주차/k-closest-points-to-origin.kt
class Solution {
    // 전체 시간복잡도: O(n log k)  (n = points.size)
    // 전체 공간복잡도: O(k)
    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        if (k == points.size) return points

        // java.util.PriorityQueue를 풀 네임으로 사용하여 import 충돌 회피
        val pq = java.util.PriorityQueue<IntArray>(Comparator { a, b ->
            val da = a[0] * a[0] + a[1] * a[1]
            val db = b[0] * b[0] + b[1] * b[1]
            // db - da 순으로 정렬 -> 최대 힙처럼 동작 (거리가 큰 것이 우선)
            db.compareTo(da)
        })

        // n개의 점을 보며 최대 k개만 유지: O(n log k)
        for (p in points) {
            pq.offer(p)
            if (pq.size > k) pq.poll() // 가장 먼 점 제거: O(log k)
        }

        // 결과 구성: O(k)
        val res = Array(k) { IntArray(2) }
        var idx = 0
        while (idx < k) {
            val arr = pq.poll()
            // poll()은 null 가능성이 있으나 pq.size == k 보장하므로 안전
            res[idx++] = arr
        }
        return res
    }
}
